### DISCUSSION TITLE
Custom select policy in R is less efficient than any default one

### DISCUSSION THREAD

**Discussion by ValerCheck** [r-simmer/simmer#242] [Q&A] [answered]
I have implemented some custom selection policy with R code and have started simulation in order to compare the default policies and my own, however from the analysis I have figured out that my R-policy is much less effective than even a "random" policy from default simmer code.

In order to get an adequate comparison results I have 2 options:
1. Rewrite 'random', 'shortest-queue' and 'round-robin' policies with R to have consistent approach to selection of resources.
2. Implement my 'custom' policy with C code and integrate it with simmer (but I couldn't find any docs on either how to implement it or how to integrate it)

Do you know how to overcome such development peculiarities in order to get adequate comparison results due to chasm between performance of R code and C code?

**Comment by Enchufa2** [comment, marked as answer]
Yes, `simmer` implements a set of common policies in C++, but provides an R interface for the general case. Built-in policies are in the internal namespace, because they are not meant to be exposed, and that's why there's no documentation about it. They're here:

https://github.com/r-simmer/simmer/blob/f58ad54d89092cc6fb8af50d400d2947cda9056b/inst/include/simmer/activity/resource_internal.h#L60-L169

Before suggesting anything, I have two questions:

- What's your purpose? Are you interested in (a) the results of the simulation applying several policies (i.e., waiting time in queues...), or (b) pure benchmarking (i.e., execution time)?
- Is your custom policy very specific or general enough to be of interest to a wide range of `simmer` users?

**Reply by ValerCheck** [reply]
If to say very general, I need to investigate the behavior of the system under the application of different policies including common simmer policies. So, I have created one simulation scenario, which I run in parallel for each policy (the only thing which is changing among simulations is a policy). In the end I compare for example cumulative mean wait time of incoming tasks with increasing number of tasks during the simulation. For example as on image.

![image](https://user-images.githubusercontent.com/10292255/110996048-02c02280-8384-11eb-859b-5ceead63cce4.png)

I wouldn't say that my specific policy can make great interest to simmer users in terms of queuing models, because I am trying to consider a bit wider range of indicators specific to my domain area. I use simmer as a tool to investigate the behavior of IT infrastructure and trying to extend details which I put into simulation models to reflect real object.

**Reply by ValerCheck** [reply]
However, I don't know, maybe my research based on `simmer` can be interesting for some part of users :)
I have described a lot in my previous comment, so to be short with regard to your question I believe that I need (a) but I need to collect results from independent simulations, so if I understand you correctly, I am not going to mix selection policies at this moment. However, I also need to investigate execution time. I need to be sure that overall processing time of all tasks is going to be similar to 'shortest-queue' policy or better.

**Reply by Enchufa2** [reply]
My point being that, if you are investigating queueing policies and not doing algorithm benchmarking, what if one particular simulation takes longer to finish? Does it matter? Execution time does not affect simulation time, does not affect the results. I'd say: leave it running a bit longer, collect the results, and go on; don't waste time optimizing if that's not your primary goal. :) If, on the contrary, the simulation takes way too long (days?, weeks?), then I'd look into optimizing that.

Regarding my second question, if your policy is general enough and only depends on the status of a list of resources, then we could look into including it in `simmer` as another built-in policy.

**Reply by ValerCheck** [reply]
Ok, let my try to paraphrase my issue :)

I have simulation which consist of 1 scheduler (who decides how to choose server via policy) and 50 servers and task generator which generate 160 tasks.

I have run this simulation 4 times with 4 different policies (shortest-queue, round-robin, random, custom). and depending on the policy those 50 servers have different total processing time. Most obvious that shortest-queue is going to be the best one, round-robin on the second place for the same simulation, and random is the worst one.

When I introduce my own policy, whose algorithm is very similar to shortest-queue but has small modification, I expect this policy to perform not worse than random, however when I compare all of them I see this:

![image](https://user-images.githubusercontent.com/10292255/111002309-e1fcca80-838d-11eb-866c-8aeaa1a5ff7d.png)

For some reason my a bit modified shortest-queue algorithm is around 6 times worse than random, however I am not selecting servers by random. Even if I remove my modification and keep only R-implemented shortest-queue part I am obtaining a similar results. If you want, I can make comparison of `simmer`'s shortest-queue and custom R-implemented shortest-queue.

P.S.
I don't think this algorithm worth adding to simmer's standard set, it is just a testing of hypothesis that my proposed approach is going to work in real domain-specific object, so it is not a general policy.

**Reply by Enchufa2** [reply]
What is "mean time to finish" here? Can you share your code, or at least a simpler example?

**Reply by ValerCheck** [reply]
Oh, maybe it was a bad idea to display it as mean. Actually, it is total time to finish arrived tasks. So, if you take for example 100 absolutely same tasks (generation time and wait timeout on scheduler, and process timeout are same for all simulations), then depending on the injected policy overall time to finish those tasks is as on figure.

**Reply by Enchufa2** [reply]
But this is simulated time reported by `get_mon_arrivals`, right? Not real time measured externally. Implementing your policy in R or C doesn't impact its performance in simulated time. If you are obtaining a non-expected performance is because there's some error in your logic, which is making arrivals e.g. taking the same resource over and over. If you can post a minimal example of your issue, I could take a look.

**Reply by ValerCheck** [reply]
Yes, that is taken from `get_mon_arrivals` and chunked into periods (let say per hour analysis). Give me some time to publish it on github, then I will attach the code. Honestly, I have also started doubting in my logic :D

**Reply by ValerCheck** [reply]
You were right, I have implemented `shortest-queue` policy in R and used it as custom, results are the same for C++ `shortest-queue` and R `shortest-queue`. There is a bottleneck in the logic of my algorithm :(

**Reply by Enchufa2** [reply]
Let me know when it's available, and I'll take a look.
